pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,357 @@
+from functools import cached_property
+import numbers
+import weakref
+
+from ..enlisting import Enlist
+from .optimization_problem import MinimizationProblem
+from .optimization_solver import OptimizationSolver
+
+
+import numpy as np
+try:
+    import petsc4py.PETSc as PETSc
+except ModuleNotFoundError:
+    PETSc = None
+
+__all__ = \
+    [
+        "TAOSolver"
+    ]
+
+
+class PETScVecInterface:
```

Comment:
This is missing docstrings which makes it really hard for me to understand.

Reply:
Documentation added. I've done my best to follow pyadjoint documentation conventions.